First Auth0 Prototype for Measurement Dashboards

chutten opened the pull request:
By all means give it a try. Head to https://chutten.github.io/telemetry-dashboard/ and get into the Measurement Dashboard. There's a little login button at the top-right. It should redirect you to my auth0 for auth 

user: user2@domain.com
pass: U$erdomain

If you successfully authenticate the login button should disappear and your access token will be at `window.Telemetry.AuthorizationToken`

Limitations: 
* Each navigation needs new auth
  * This includes refreshes and new tabs in addition to navigating between evo and dist
  * Auth0 is smart enough to auth you almost silently, so subsequent logins are just a matter of clicking the login button. No credentials. This may be low-friction enough for MVP, or may not.
* No UI to let users know if they've auth'd or not.
  * The entire design is in need of review, so this isn't much surprise
* Doesn't transmit auth to aggregates.tmo
  * aggregates.tmo actually has some CORS on it saying we can't send any but the most standard of headers. `Authorization` isn't a "most standard" so I had to comment it out.
* No handling for token expiry
  * Tokens are configured to be valid for 2 hours, but I've not implemented any way of renewing it beforehand or handling a case where auth goes wrong.
  * User sessions are short enough and Auth0 authentication is seamless enough that this might be acceptable for MVP.

chutten committed:
First Auth0 Prototype for Measurement Dashboards

chutten commented:
@wlach @mreid-moz  @fbertsch You may find this interesting

fbertsch commented on new-pipeline/src/auth.js:
```diff
@@ -0,0 +1,102 @@
+  return loginUrl.toString();
+}
+
+function createLoginButton() {
```
This should probably be moved to html/css, no?

chutten replied:
It can be. Creating it in JS is an artefact from when it was a disappearing login button.

chutten committed:
Send the Authorization header, use staging

chutten committed:
Ensure logout clears mdv1 app state

chutten committed:
Telemetry.getVersions should return [] if from and to have different channels.

Telemetry.getVersions will lexicographically filter candidate channel+version
pairs based on fromVersion and toVersion.

If fromVersion is the default min_channel_version (which when null is taken to
be the 3rd-most-recent nightly version) it works acceptably so long as the
max_channel_version starts with a letter in the alphabet before "n".

With release-channel aggregates coming back, "r" isnt' before "n" so this bug
becomes a problem.

chutten committed:
Factor out common state-saving code

chutten committed:
Support auto-login when trying to load app state with release channels.

When being shared a link pointing at release data, when moving from evo to dist
or vice versa, or when simply reloading the page a user will no longer be
authorized to view release data but their state will have release-channel state
in it.

In this case we'll pop out to the authorization flow automatically as though
authorization had failed.

In most cases I expect this to immediately redirect back to the app in-state.

chutten committed:
403 is Authorization failure, 401 is Authentication failure

chutten committed:
Authenticate against mozilla's auth0 instance

fbertsch commented on new-pipeline/src/auth.js:
```diff
@@ -0,0 +1,118 @@
+let gRequestState;
+
+/* Copied with implicit license from Auth0 docs: randomString */
+function randomString(length) {
```
There's really no available library function?

chutten replied:
Even if there were, we already have a problem of too many JS files on load.

fbertsch commented on new-pipeline/src/auth.js:
```diff
@@ -0,0 +1,118 @@
+  return logoutUrl.toString();
+}
+
+function displayAuthButton(buttonText, buttonUrl) {
```
nit: shouldn't this be defined in the HTML?

chutten replied:
Actually, upon reflection I'd like to leave it. Messing with individual text nodes to change the button text and href... we can move it to a follow-up

fbertsch reviewed (approved):
Just a few nits. Let's test this
